tests/console/vmstat.pm

```diff
@@ -0,0 +1,92 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Package: vmstat
+# Summary: Test vmstat installation and verify vmstat and compare the output with/without load.
+# - Install procps
+# - Run vmstat with options
+# - Run vmstat without load and Read output
+# - Run vmstat with load and Read output
+# - Compare the output of vmstat with and without load
+# Maintainer: QE Core <qe-core@suse.de> 
+
+use base 'opensusebasetest';
+use testapi;
+use utils;
+use strict;
+use warnings;
+use POSIX;
+use version_utils;
+use registration qw(add_suseconnect_product get_addon_fullname);
+
+use List::Util qw(min);
+use console::vmstat_utils;
+
+sub run {
+
+    my ($self) = @_;
+    my ($vm, $cpu);
+    $self->select_serial_terminal;
+    zypper_call('in procps');
+    assert_script_run("rpm -q procps");
+    validate_script_output("vmstat",
+        qr/(procs\s-+memory-+\s-+swap-+\s-+io-+\s-+system-+\s-+cpu-+).*(\s+|\d+)+/s);
+
+    # Extend vmstat utility. Verifying the vmstat with options.
+    validate_script_output("vmstat -t",
+        qr/(procs\s-+memory-+\s-+swap-+\s-+io-+\s-+system-+\s-+cpu-+\s-+timestamp-+).*(\s+|\d+)+/s);
+    validate_script_output("vmstat -a",
+        qr/(procs\s-+memory-+\s-+swap-+\s-+io-+\s-+system-+\s-+cpu-+).*(\s+.*\s+inact\s+active\s*.*).*(\s+|\d+)+/s);
+    validate_script_output("vmstat -f", qr/(\d+\s+forks)/s);
+
+    #Running vmstat under light load and dumping the output to a file. Calculate and collect the minimum idle memory from the output.
+    assert_script_run("sync; echo 3 > /proc/sys/vm/drop_caches");
+    assert_script_run("vmstat 2 10 > /tmp/output_light.log");
+    assert_script_run("cat /tmp/output_light.log");
+    my ($mem_lite_load_ref, $cpu_lite_load_ref) = read_memory_cpu("/tmp/output_light.log");
+    my @mem_lite_load     = @$mem_lite_load_ref;
+    my @cpu_lite_load     = @$cpu_lite_load_ref;
+    my $min_mem_lightload = min(@mem_lite_load);
+    my $min_cpu_lightload = min(@cpu_lite_load);
+
+    # Increasing the load in the system using package stress-ng.
+    zypper_call('in stress-ng');
+    assert_script_run("rpm -q stress-ng");
+    my $total_cpu = script_output("cat /proc/cpuinfo | grep processor | wc -l");
+    my $total_mem = script_output("free -h | awk '/Mem\:/ { print \$2 }'");
+    if ($total_cpu == 1) { $cpu = $total_cpu; } else { $cpu = floor($total_cpu * 0.75); }
+    $vm = 50;
+    my $stressng_pkg = "nohup stress-ng --cpu $cpu --vm $vm --vm-bytes 256m --timeout 60s&";
+    enter_cmd "$stressng_pkg";
+    sleep 10;
```

Comment:
You are using `sleep()` or are simulating a sleep-like functionality. This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. You *can* use sleep but there are only very limited cases where is the best solution (e.g. yield threads in low-level system code). Here you have the classical problem in software design of [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science)) . sleep should not be used in place of proper synchronization methods. https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/ is one of multiple references that explains this: "The original problem is likely a timing/synchronization issue, ignoring it by hiding it with Thread.Sleep is only going to delay the problem and make it occur in random, hard to reproduce ways." And if the problem does not occur at least time is wasted. https://testing.googleblog.com/2021/03/test-flakiness-one-of-main-challenges.html states that as well: "Do NOT add arbitrary delays as these can become flaky again over time and slow down the test unnecessarily."

We are waiting because some condition changes some time ... keyword(s) is/are *some time*. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time. For openQA tests in general just try to do what you as a human user would also do. You would not look on your watch waiting for exactly 100s before you look at the screen again, right? :)

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

Reply:
I am collecting vmstat output after the 10s delay of  stress-ng run, which is running in the background and so kept sleep of 10s. Later I will compare the vmstat sample with light load.

Reply:
Well, as I stated, you are waiting for *something*, so wait for that, not sleep 10

Reply:
Let's keep the sleep for the time being, the extra work needed to check whether the load did really increase due to stress-ng will be catched up anyway further ahead.

Reply:
we will keep the sleep.